src/frequenz/sdk/microgrid/component_data.py

```diff
@@ -6,240 +6,243 @@
 License
 MIT
 """
+from __future__ import annotations
 
-from abc import ABC
+from abc import ABC, abstractmethod
+from dataclasses import dataclass, field
 from datetime import datetime
-from typing import Tuple
+from typing import Optional, Tuple
 
 import frequenz.api.microgrid.microgrid_pb2 as microgrid_pb
 import pytz
 
 from .component_states import EVChargerCableState
 
 
-class _BaseComponentData(ABC):
-    """A private base class for strongly typed component data classes."""
+@dataclass(frozen=True)
+class ComponentData(ABC):
+    """A private base class for strongly typed component data classes.
 
-    _raw: microgrid_pb.ComponentData
+    Attributes:
+        component_id: the ID identifying this component in the microgrid
+        timestamp: the timestamp of when the data was measured.
+    """
 
-    @property
-    def component_id(self) -> int:
-        """Return the component id of the message.
+    component_id: int
+    timestamp: datetime
+    _raw: Optional[microgrid_pb.ComponentData] = field(default=None, init=False)
```

Comment:
Please add a comment here explaining why we have the `_raw` data here, why is it optional and why is `init=False` being used.

Reply:
Also thinking more about this, I think `_raw` should be a public member (`raw`). Even when it's a bit hidden, it is intended to be used by users, is not something internal.

Reply:
Makes sense. Changed to a public member.